crates/core/tedge_agent/src/tedge_operation_converter/actor.rs

```diff
@@ -58,93 +69,124 @@ impl Actor for TedgeOperationConverterActor {
-                    cmd_id,
-                },
-            )) => match SoftwareListCommand::try_from(target, cmd_id, message.payload_bytes()) {
-                Ok(Some(cmd)) => {
-                    self.software_sender.send(cmd.into()).await?;
-                }
-                Ok(None) => {
-                    // The command has been fully processed
-                }
-                Err(err) => error!("Incorrect software_list request payload: {err}"),
-            },
-
-            // Software Update
-            Ok((
-                target,
-                Channel::Command {
-                    operation: OperationType::SoftwareUpdate,
-                    cmd_id,
-                },
-            )) => match SoftwareUpdateCommand::try_from(target, cmd_id, message.payload_bytes()) {
-                Ok(Some(cmd)) => {
-                    self.software_sender.send(cmd.into()).await?;
-                }
-                Ok(None) => {
-                    // The command has been fully processed
+        let (target, operation, cmd_id) = match self.mqtt_schema.entity_channel_of(&message.topic) {
+            Ok((target, Channel::Command { operation, cmd_id })) => (target, operation, cmd_id),
+
+            _ => {
+                log::error!("Unknown command channel: {}", message.topic.name);
+                return Ok(());
+            }
+        };
+
+        match self
+            .workflows
+            .get_workflow_current_action(&operation, &message)
+        {
+            Ok(None) | Ok(Some((_, OperationAction::Clear))) => {
+                // The command has been fully processed
+                Ok(())
+            }
+            Ok(Some((state, OperationAction::MoveTo(next_step)))) => {
+                let new_state = state.move_to(next_step);
+                self.publish_command_state(operation, cmd_id, new_state)
+                    .await
+            }
+            Ok(Some((state, OperationAction::BuiltIn))) => {
+                let step = &state.status;
+                info!("Processing {operation} operation {step} step");
+                self.process_internal_operation(target, operation, cmd_id, message)
+                    .await
+            }
+            Ok(Some((state, OperationAction::Delegate(participant)))) => {
+                let step = &state.status;
+                info!("Delegating {operation} operation {step} step to: {participant}");
+                // TODO fail the operation on timeout
+                Ok(())
+            }
+            Ok(Some((state, OperationAction::Script(script)))) => {
+                let step = &state.status;
+                info!("Processing {operation} operation {step} step with script: {script}");
+                if let Ok(mut command) = Execute::try_new(&script) {
+                    command.args = state.inject_parameters(&command.args);
+                    let output = self.script_runner.await_response(command).await?;
```

Comment:
> Once we start supporting child device routines also via workflows, this blocking could be a problem

What do you mean? Workflow execution is supervised by `tedge-agent`, possibly on a child-device.

> For now, this might be a blessing in disguise, allowing serial execution of operations on the main device.

Not an accident.